File: shared/WaiSayKit/Sources/WaiSayKit/Monitoring/SentryHelper.swift
import CryptoKit
import Foundation
import Sentry

public enum SentryHelper {
    public static func start(dsn: String, debug: Bool = false) {
        SentrySDK.start { options in
            options.dsn = dsn
            options.debug = debug
            options.tracesSampleRate = 0.1
            options.profilesSampleRate = 0.1
            options.enableAutoSessionTracking = true
            options.enableCaptureFailedRequests = true
            #if canImport(UIKit)
            options.attachScreenshot = false
            #endif
            options.sendDefaultPii = false
            #if DEBUG
            options.environment = "development"
            #else
            options.environment = "production"
            #endif
        }
    }

    public static func captureError(_ error: Error, extras: [String: Any]? = nil) {
        SentrySDK.capture(error: error) { scope in
            if let extras {
                for (key, value) in sanitizeDictionary(extras) {
                    scope.setExtra(value: value, key: key)
                }
            }
        }
    }

    public static func addBreadcrumb(
        category: String,
        message: String,
        level: SentryLevel = .info,
        data: [String: Any]? = nil
    ) {
        let crumb = Breadcrumb(level: level, category: category)
        crumb.message = sanitizeString(message, key: "message")
        if let data {
            crumb.data = sanitizeDictionary(data)
        }
        SentrySDK.addBreadcrumb(crumb)
    }

    public static func setUser(id: String) {
        let user = Sentry.User()
        user.userId = id
        SentrySDK.setUser(user)
    }

    public static func clearUser() {
        SentrySDK.setUser(nil)
    }

    static func sanitizeDictionary(_ values: [String: Any]) -> [String: Any] {
        var sanitized: [String: Any] = [:]
        for (key, value) in values {
            sanitized[key] = sanitizeValue(value, key: key)
        }
        return sanitized
    }

    static func sanitizeValue(_ value: Any, key: String? = nil) -> Any {
        if let dictionary = value as? [String: Any] {
            return sanitizeDictionary(dictionary)
        }

        if let array = value as? [Any] {
            return array.map { sanitizeValue($0, key: key) }
        }

        if let string = value as? String {
            return sanitizeString(string, key: key)
        }

        if let data = value as? Data {
            return "<bytes:\(data.count)>"
        }

        return value
    }

    static func sanitizeString(_ value: String, key: String? = nil) -> String {
        let normalizedKey = key?.lowercased() ?? ""
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)

        if containsAny(normalizedKey, fragments: ["token", "password", "secret", "authorization", "cookie"]) {
            return "[redacted-secret]"
        }
        if containsAny(normalizedKey, fragments: ["email"]) {
            return "[redacted-email:\(fingerprint(trimmed.lowercased()))]"
        }
        if containsAny(normalizedKey, fragments: ["filename", "file_name"]) {
            let ext = URL(fileURLWithPath: trimmed).pathExtension.lowercased()
            return "[redacted-filename:ext=\(ext.isEmpty ? "none" : ext):sha=\(fingerprint(trimmed)):len=\(trimmed.count)]"
        }
        if containsAny(normalizedKey, fragments: ["query", "question", "text", "content", "transcript", "body", "html", "prompt", "reason", "error", "detail", "description"]) {
            return "[redacted-text:len=\(trimmed.count):sha=\(fingerprint(trimmed))]"
        }

        return redactEmbeddedSecrets(in: value)
    }

    private static func containsAny(_ key: String, fragments: [String]) -> Bool {
        fragments.contains { key.contains($0) }
    }

    private static func fingerprint(_ value: String) -> String {
        guard !value.isEmpty else { return "-" }
        let digest = SHA256.hash(data: Data(value.utf8))
        return digest.compactMap { String(format: "%02x", $0) }.joined().prefix(12).description
    }

    private static func redactEmbeddedSecrets(in value: String) -> String {
        guard let emailRegex = try? NSRegularExpression(
            pattern: "[A-Z0-9._%+-]+@[A-Z0-9.-]+\\.[A-Z]{2,}",
            options: [.caseInsensitive]
        ) else {
            return value
        }

        let range = NSRange(value.startIndex..<value.endIndex, in: value)
        var result = value

        for match in emailRegex.matches(in: value, options: [], range: range).reversed() {
            guard let matchRange = Range(match.range, in: result) else { continue }
            let email = String(result[matchRange]).lowercased()
            result.replaceSubrange(
                matchRange,
                with: "[redacted-email:\(fingerprint(email))]"
            )
        }

        return result
    }
}

File: android/app/src/main/java/is/waiwai/computer/monitoring/SentryHelper.kt
package `is`.waiwai.computer.monitoring

import android.content.Context
import io.sentry.Breadcrumb
import io.sentry.Sentry
import io.sentry.SentryLevel
import io.sentry.android.core.SentryAndroid
import java.security.MessageDigest
import java.util.concurrent.ConcurrentHashMap

object SentryHelper {
    private val lastBreadcrumbAt = ConcurrentHashMap<String, Long>()

    fun start(context: Context, dsn: String) {
        if (dsn.isBlank()) return
        // Android MUST use SentryAndroid.init; the core Sentry.init throws
        // "You are running Android" and crashes Application.onCreate.
        SentryAndroid.init(context) { options ->
            options.dsn = dsn
            options.beforeBreadcrumb = io.sentry.SentryOptions.BeforeBreadcrumbCallback { breadcrumb, _ ->
                sanitizeBreadcrumb(breadcrumb)
            }
            options.beforeSend = io.sentry.SentryOptions.BeforeSendCallback { event, _ ->
                event.user = null
                event.tags?.keys?.toList().orEmpty().forEach { key ->
                    event.setTag(key, sanitizeString(event.tags?.get(key).orEmpty(), key))
                }
                event
            }
        }
    }

    fun setUser(id: String) {
        Sentry.setUser(io.sentry.protocol.User().apply { this.id = id })
    }

    fun clearUser() {
        Sentry.setUser(null)
    }

    fun captureError(t: Throwable, extras: Map<String, Any> = emptyMap()) {
        Sentry.captureException(t) { scope ->
            sanitizeMap(extras).forEach { (key, value) ->
                scope.setExtra(key, value.toString())
            }
        }
    }

    fun addBreadcrumb(
        category: String,
        message: String,
        level: SentryLevel = SentryLevel.INFO,
        data: Map<String, Any> = emptyMap(),
    ) {
        val now = System.currentTimeMillis()
        val previous = lastBreadcrumbAt[category]
        if (previous != null && now - previous < BREADCRUMB_THROTTLE_MS) {
            return
        }
        lastBreadcrumbAt[category] = now
        val breadcrumb = Breadcrumb().apply {
            this.category = category
            this.message = sanitizeString(message, "message")
            this.level = level
            sanitizeMap(data).forEach { (key, value) ->
                setData(key, value)
            }
        }
        Sentry.addBreadcrumb(breadcrumb)
    }

    private fun sanitizeBreadcrumb(breadcrumb: Breadcrumb?): Breadcrumb? {
        breadcrumb ?: return null
        breadcrumb.message = sanitizeString(breadcrumb.message.orEmpty(), "message")
        val data = breadcrumb.data ?: return breadcrumb
        data.keys.toList().forEach { key ->
            breadcrumb.setData(key, sanitizeValue(key, data[key] ?: ""))
        }
        return breadcrumb
    }

    private fun sanitizeMap(values: Map<String, Any>): MutableMap<String, Any> {
        return values.mapValuesTo(mutableMapOf()) { (key, value) ->
            sanitizeValue(key, value)
        }
    }

    private fun sanitizeValue(key: String, value: Any): Any {
        return when (value) {
            is String -> sanitizeString(value, key)
            is ByteArray -> "<bytes:${value.size}>"
            is Map<*, *> -> value.entries.associate { (nestedKey, nestedValue) ->
                nestedKey.toString() to sanitizeValue(nestedKey.toString(), nestedValue ?: "")
            }
            is Iterable<*> -> value.map { sanitizeValue(key, it ?: "") }
            else -> value
        }
    }

    private fun sanitizeString(value: String, key: String): String {
        val normalizedKey = key.lowercase()
        val trimmed = value.trim()
        return when {
            SECRET_KEYS.any(normalizedKey::contains) -> "[REDACTED]"
            EMAIL_KEYS.any(normalizedKey::contains) -> "[REDACTED:${fingerprint(trimmed.lowercase())}]"
            FILE_KEYS.any(normalizedKey::contains) -> "[REDACTED:${fingerprint(trimmed)}]"
            TEXT_KEYS.any(normalizedKey::contains) -> "[REDACTED:${trimmed.length}:${fingerprint(trimmed)}]"
            else -> EMAIL_REGEX.replace(trimmed) { matchResult ->
                "[REDACTED:${fingerprint(matchResult.value.lowercase())}]"
            }
        }
    }

    private fun fingerprint(value: String): String {
        if (value.isEmpty()) return "-"
        return MessageDigest.getInstance("SHA-256")
            .digest(value.toByteArray())
            .joinToString("") { byte -> "%02x".format(byte) }
            .take(12)
    }

    private val SECRET_KEYS = listOf("token", "password", "secret", "authorization", "cookie")
    private val EMAIL_KEYS = listOf("email")
    private val FILE_KEYS = listOf("filename", "file_name", "title")
    private val TEXT_KEYS = listOf("transcript", "query", "question", "text", "content", "body", "detail")
    private val EMAIL_REGEX = Regex("[A-Z0-9._%+-]+@[A-Z0-9.-]+\\.[A-Z]{2,}", RegexOption.IGNORE_CASE)
    private const val BREADCRUMB_THROTTLE_MS = 5_000L
}
